File: dotnet-soa/ProductService/Domain/Interfaces/IProductService.cs
using ProductService.Domain.Models;

namespace ProductService.Domain.Interfaces;

/// <summary>
/// Product service interface following Interface Segregation Principle (ISP).
/// 
/// SOA Principle: Service abstraction - consumers interact with the service
/// through a well-defined contract without knowing implementation details.
/// </summary>
public interface IProductService
{
    /// <summary>Creates a new product and publishes a product.created event.</summary>
    Task<Product> CreateProductAsync(string name, string? description, decimal price, string? category);

    /// <summary>Retrieves all products from the catalog.</summary>
    Task<IEnumerable<Product>> GetAllProductsAsync();

    /// <summary>Retrieves a product by its unique identifier.</summary>
    Task<Product?> GetProductByIdAsync(Guid id);

    /// <summary>Deletes a product from the catalog.</summary>
    Task<bool> DeleteProductAsync(Guid id);
}

File: nodejs-soa/product-service/src/domain/service/IProductService.ts
// =============================================================================
// Product Service Interface (Contract)
// =============================================================================
// This interface defines the contract for the Product Service.
// 
// SOLID Principles Applied:
// - ISP (Interface Segregation): This interface is focused only on product operations
// - DIP (Dependency Inversion): Controllers depend on this abstraction, not implementations
// - LSP (Liskov Substitution): Any implementation can be substituted without breaking clients
// =============================================================================

import { Product, CreateProductData, UpdateProductData } from '../model/Product';

/**
 * IProductService - Service contract for product catalog operations.
 * 
 * In SOA, service interfaces define the contract that the service exposes.
 * This enables loose coupling between the service layer and its consumers.
 */
export interface IProductService {
  /**
   * Retrieve all active products from the catalog.
   */
  findAll(): Promise<Product[]>;

  /**
   * Find a specific product by its unique identifier.
   * @param id - Product UUID
   * @returns Product if found, null otherwise
   */
  findById(id: string): Promise<Product | null>;

  /**
   * Create a new product in the catalog.
   * Publishes a 'product.created' event to the ESB.
   * @param data - Product creation data
   * @returns The created product
   */
  create(data: CreateProductData): Promise<Product>;

  /**
   * Update an existing product.
   * @param id - Product UUID
   * @param data - Fields to update
   * @returns Updated product if found, null otherwise
   */
  update(id: string, data: UpdateProductData): Promise<Product | null>;

  /**
   * Soft-delete a product (marks as inactive).
   * @param id - Product UUID
   * @returns true if deleted, false if not found
   */
  delete(id: string): Promise<boolean>;
}
